Cmd+Shift+Up/Down and Cmd+Opt+Up/Down shouldn't scroll a web page

anurag-23 opened the pull request:
#### 1189a768e05bf12f04b0b0129fa4db48cc30d7c8
<pre>
Cmd+Shift+Up/Down and Cmd+Opt+Up/Down shouldn&apos;t scroll a web page
<a href="https://bugs.webkit.org/show_bug.cgi?id=282852">https://bugs.webkit.org/show_bug.cgi?id=282852</a>
<a href="https://rdar.apple.com/139537723">rdar://139537723</a>

Reviewed by Abrar Rahman Protyasha and Wenson Hsieh.

The designated shortcut for &quot;Scroll to Top/Bottom&quot; is Cmd+Up/Down Arrow.
Cmd+Shift/Opt/Ctrl+Up/Down are not valid key combinations for web page
scrolling, and might be designated shortcuts for other actions instead.

We already handled the case where Shift is pressed in 287161@main. This patch
ensures that while handling an Up/Down keyboard event with the Command key
pressed, we return a `Document` scroll granularity from `scrollGranularityForKeyboardEvent()`
only if no other modifier is pressed at the same time. If another modifier is
found to be pressed, we don&apos;t return a valid granularity so that no scroll
occurs.

Since `shouldBeTrue()`/`shouldBeEqual()` calls in the layout tests lead to debug
output that affects the scroll position, these need to be called with the
`quiet` parameter set to true.

* LayoutTests/fast/scrolling/ios/scroll-to-beginning-and-end-of-document-expected.txt:
* LayoutTests/fast/scrolling/ios/scroll-to-beginning-and-end-of-document.html:
* LayoutTests/fast/scrolling/mac/scroll-to-beginning-and-end-of-document-expected.txt: Copied from LayoutTests/fast/scrolling/ios/scroll-to-beginning-and-end-of-document-expected.txt.
* LayoutTests/fast/scrolling/mac/scroll-to-beginning-and-end-of-document.html: Added.
* Source/WebCore/platform/KeyboardScrollingAnimator.cpp:
(WebCore::scrollGranularityForKeyboardEvent):

Canonical link: <a href="https://commits.webkit.org/287437@main">https://commits.webkit.org/287437@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ceb23041c796bf3a0713cc9b5e79f1f062352e7f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79507 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58516 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32884 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84086 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30600 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81641 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67609 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6794 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62171 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20035 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82578 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/52231 "Exiting early after 10 failures. 10 tests run. 1 failures") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72434 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42481 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49579 "Exiting early after 10 failures. 10 tests run. 1 failures") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26597 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29016 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70698 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27055 "Found 1 new test failure: fast/editing/recursive-reapply-edit-command-crash.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85497 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6773 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4712 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70422 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6938 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68296 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69665 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17397 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13694 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12587 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6725 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12412 "") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6621 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10103 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8420 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (approved):
Do you mind rebasing your patch against main? [287161@main](https://commits.webkit.org/287161@main) touched the same codepath and I'd like to verify we don't introduce an oddity when these two commits coexist.

whsieh reviewed (approved):
It looks like the EWS failure here is specific to iOS 17, and the test is otherwise appropriately marked as failing.

Commit:
Cmd+Shift+Up/Down and Cmd+Opt+Up/Down shouldn't scroll a web page
https://bugs.webkit.org/show_bug.cgi?id=282852
rdar://139537723

Reviewed by Abrar Rahman Protyasha and Wenson Hsieh.

The designated shortcut for "Scroll to Top/Bottom" is Cmd+Up/Down Arrow.
Cmd+Shift/Opt/Ctrl+Up/Down are not valid key combinations for web page
scrolling, and might be designated shortcuts for other actions instead.

We already handled the case where Shift is pressed in 287161@main. This patch
ensures that while handling an Up/Down keyboard event with the Command key
pressed, we return a `Document` scroll granularity from `scrollGranularityForKeyboardEvent()`
only if no other modifier is pressed at the same time. If another modifier is
found to be pressed, we don't return a valid granularity so that no scroll
occurs.

Since `shouldBeTrue()`/`shouldBeEqual()` calls in the layout tests lead to debug
output that affects the scroll position, these need to be called with the
`quiet` parameter set to true.

* LayoutTests/fast/scrolling/ios/scroll-to-beginning-and-end-of-document-expected.txt:
* LayoutTests/fast/scrolling/ios/scroll-to-beginning-and-end-of-document.html:
* LayoutTests/fast/scrolling/mac/scroll-to-beginning-and-end-of-document-expected.txt: Copied from LayoutTests/fast/scrolling/ios/scroll-to-beginning-and-end-of-document-expected.txt.
* LayoutTests/fast/scrolling/mac/scroll-to-beginning-and-end-of-document.html: Added.
* Source/WebCore/platform/KeyboardScrollingAnimator.cpp:
(WebCore::scrollGranularityForKeyboardEvent):

Canonical link: https://commits.webkit.org/287437@main
